src/vcpkg/paragraphs.cpp

```diff
@@ -108,51 +33,100 @@ namespace vcpkg
-    {
-        auto maybe_field = remove_field(&fields, fieldname);
-        if (const auto field = maybe_field.get())
-            out = std::move(*field);
-        else
-            missing_fields.emplace_back(fieldname.data());
-    }
-    void ParagraphParser::optional_field(StringLiteral fieldname, std::pair<std::string&, TextRowCol&> out)
+    std::string ParagraphParser::optional_field(StringLiteral fieldname, TextRowCol& position)
     {
         auto maybe_field = remove_field(&fields, fieldname);
-        if (auto field = maybe_field.get()) out = std::move(*field);
-    }
-    void ParagraphParser::required_field(StringLiteral fieldname, std::string& out)
-    {
-        TextRowCol ignore;
-        required_field(fieldname, {out, ignore});
+        if (auto field = maybe_field.get())
+        {
+            position = field->second;
+            return std::move(field->first);
+        }
+
+        return std::string();
     }
+
     std::string ParagraphParser::optional_field(StringLiteral fieldname)
     {
-        std::string out;
         TextRowCol ignore;
-        optional_field(fieldname, {out, ignore});
-        return out;
+        return optional_field(fieldname, ignore);
     }
+
     std::string ParagraphParser::required_field(StringLiteral fieldname)
     {
-        std::string out;
-        TextRowCol ignore;
-        required_field(fieldname, {out, ignore});
-        return out;
+        auto maybe_field = remove_field(&fields, fieldname);
+        if (const auto field = maybe_field.get())
+        {
+            return std::move(field->first);
+        }
+
+        errors.emplace_back(LocalizedString::from_raw(origin)
+                                .append_raw(": ")
+                                .append_raw(ErrorPrefix)
+                                .append(msgMissingRequiredField2, msg::json_field = fieldname));
+        return std::string();
+    }
+
+    void ParagraphParser::add_error(TextRowCol position, msg::MessageT<> error_content)
+    {
+        errors.emplace_back(LocalizedString::from_raw(origin)
+                                .append_raw(fmt::format("{}:{}: ", position.row, position.column))
+                                .append_raw(ErrorPrefix)
+                                .append(error_content));
     }
 
-    std::unique_ptr<ParseControlErrorInfo> ParagraphParser::error_info(StringView name) const
+    Optional<LocalizedString> ParagraphParser::error() const
     {
-        if (!fields.empty() || !missing_fields.empty())
+        LocalizedString result;
+        if (errors.empty())
         {
-            auto err = std::make_unique<ParseControlErrorInfo>();
-            err->name = name.to_string();
-            err->extra_fields = Util::extract_keys(fields);
-            err->missing_fields = missing_fields;
-            err->expected_types = expected_types;
-            return err;
+            if (fields.empty())
+            {
+                return nullopt;
+            }
```

Comment:
Is the split of the ifs intentional? If so, I think a comment would be nice so people don't accidently merge them.

Reply:
Yes. 

But 
```c++
if (!errors.empty()) {
   ...
} else if (fields.empty()) {
  return nullopt;
}
...
```
is maybe more obvious

Reply:
> But
> 
> ```c++
> if (!errors.empty()) {
>    ...
> } else if (fields.empty()) {
>   return nullopt;
> }
> ...
> 
> is maybe more obvious

I'm not a fan of this form because when analyzing the else block it forces the reader to negate a negation.

I'll add a comment that it's intentional